### PR TITLE
Memory efficient replay route

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

This adds a `ReplayRouteSession` that is able to replay super large routes at high frequencies, without running out of memory. It also integrates it into drop-in-ui, so that we can all use it!

How it works. The developer pushes events into a replayer, and the replayer simulates the events in real time. The current replayer will take all events from a route and hold them in memory. The new `ReplayRouteSession` will only replay `decodeMinDistance` events at a time. For example, if you create a 1000 kilometers route and set the decodeMinDistance to 10 kilometers. You can expect `ReplayRouteSession` to only hold ~10km of events in memory at a time. 

### Api explanation

This object is experimental but for the intents and purposes it will become permanent. The details under the cover are subject to change, but the top interface is going to be set in stone.

``` kotlin
// When enabled, mapbox navigation will automatically drive the active route
class ReplayRouteSession : MapboxNavigationObserver {
  // Enable route replay
  fun onAttached(mapboxNavigation)

  // Disable route replay
  fun onDetached(mapboxNavigation)

  // Get the current options
  fun getOptions(): ReplayRouteSessionOptions)

  // Change the options
  fun setOptions(options: ReplayRouteSessionOptions)
}
```

#### Known issue

The `ReplayRouteMapper` was designed to simulate events from A to B, where the starting speed and the ending speed is 0.0 meters per second. `ReplayRouteSession` is using the `ReplayRouteMapper` to simulate each section of the route, so the driver will appear to stop and speed up again. The larger the `decodeMinDistance`, the less often this will happen.

Also note that this issue is fixable but will require more time and effort. 

Here are memory approximations to help you pick `frequency` and `decodeMinDistance` values. Duration is more accurate for determining memory sizes, but duration requires more resources to calculate than distance. The parameters may be removed after the internal improvements are completed.

Screen Shot 2022-11-10 at 3 17 15 PM

### What kind of memory improvement to expect

Old formulas to determine memory requirements
- event_size ~500-300 bytes
- memory_required = (duration * frequency) * event_size

New formulas
- event_size ~500-300 bytes
- memory_required = (decodeMinDistance * frequency) * event_size

In other words: When duration becomes 40 hours. The memory requirement can be clamped to 30 minutes.

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteSession.kt [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+ *         )
+ *         .build()
+ * )
+ * ```
+ *
+ * Enable and disable the [ReplayRouteSession] with [MapboxNavigation] or [MapboxNavigationApp].
+ * The replay session will be enabled when [MapboxNavigation] is attached.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class ReplayRouteSession : MapboxNavigationObserver {
```
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteSession.kt:53]: We have users who still create `MapboxNavigaiton` using constructor. Will they be able to use this lazy mechanism without migrating to `MapboxNavigationApp`?

**Review by VysotskiVadim — COMMENTED** [review]
@kmadsen , lazy initialisation is a great idea for replay events! It happened to that I wasn't able to simulate 10Hz location updates for Gdansk-Lisbon route, app crashed with OOM

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayPolylineDecodeStream.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            temp = encodedPath[index++].code - 63 - 1
+            result += temp shl shift
+            shift += 5
+        } while (temp >= 0x1f)
+        lng += if (result and 1 != 0) (result shr 1).inv() else result shr 1
+
+        return Point.fromLngLat(lng / factor, lat / factor).also { next ->
+            current = next
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayPolylineDecodeStream.kt:73]: I think there are a few options. The other option is putting the geometry string into a file, and letting the file be the cache. We can pull from the file with an InputStream https://github.com/mapbox/mapbox-java/blob/65e6d80456c0f3aa5e95611785358e7815ccf6e9/services-geojson/src/test/resources/polyline-sfo-nyc.txt

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteMapper.kt [hunk 3]
```diff
@@ -88,7 +87,17 @@ class ReplayRouteMapper @JvmOverloads constructor(
...
+    }
+
+    /**
+     * Take a list of [Point] and map it to events that can be replayed by the [MapboxReplayer].
+     *
+     * @param points containing location coordinates to be replayed.
+     * @return [ReplayEventBase] [List]
+     */
+    fun mapPointList(points: List<Point>): List<ReplayEventBase> {
+        return replayRouteDriver.drivePointList(options, points)
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteMapper.kt:100]: All of the functions in the `ReplayRouteMapper` return a `List`. And the replayer still takes events as a list
[`fun pushEvents(events: List<ReplayEventBase>)`](https://github.com/mapbox/mapbox-navigation-android/blob/03d1fc09fd954d38384b8ee8de2528ea74872cd7/libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/MapboxReplayer.kt#L39)

So I don't think it is easier to convert back and forth at this time.

[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteMapper.kt:100]: The algorithms under the covers are still using lists, because they dedup locations on the route and they estimate the route curvatures. I agree that iterable sequence can be better, but in this case we're also sharing algorithms and moving code back and forth between mapbox java https://github.com/mapbox/mapbox-java/blob/1b867fe8e68b299f37e5591c8431df89881ca9c2/services-geojson/src/main/java/com/mapbox/geojson/utils/PolylineUtils.java#L39. Some of the decisions were based on the tools that exist in mapbox-java

It would be a big change to try and turn everything into a sequence

[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteMapper.kt:100]: Considering that we are trying to make replay "streamable", I also agree Sequence/Iterable/InputStream is where we should be going. But the `ReplayRouteMapper` is going to continue to be a `List` creator.

Take a look at `ReplayRouteDriver`, that is what we want to work on making into a `Sequence`. And then the mapper will convert the sequence into a `List` to maintain backwards compatibility.

I've been thinking about this too, so am just sharing the thoughts on this.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteSession.kt (hunk 1 above)
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteSession.kt:53]: Yeah. The `MapboxNavigationObserver` can be used without `MapboxNavigationApp`. The interface exposes the functions.

``` kotlin
interface MapboxNavigationOberver {
  fun onAttached(mapboxNavigation: MapboxNavigation)
  fun onDetached(mapboxNavigation: MapboxNavigation)
}
```

This means you can use `ReplayRouteSession` in this way. Call start and stop replay whenever you have an instance of `MapboxNavigation`. But again, this is NOT RECOMMENDED because we want to share this object with android auto.

``` kotlin
val replayRouteSession = ReplayRouteSession()

fun startReplay() {
  replayRouteSession.onAttached(mapboxNavigation)(
}

fun stopReplay() {
  replayRouteSession.onDetached(mapboxNavigation)(
}
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayPolylineDecodeStream.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayPolylineDecodeStream.kt:73]: Right now this `ReplayPolylineDecodeStream` is internal and considered temporary. I still want to decode it as a Iterable stream because that makes the usage easier. But if the geometry is stored in a file, cache, or memory. that doesn't matter.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteSession.kt (hunk 1 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteSession.kt:53]: Thanks!

**Review by VysotskiVadim — APPROVED** [review]
I tested on the route Gdańsk(Poland)-Ho Chi Minh City(Vietnam) with 10Hz location update. Java heap don't grow more the 0.2 Gb with the new solution, but test app crashes with OOM on main branch.

Well done! 💪 
Let's merge and iterate improving this solution 🚀 
Improvements I recommend:
* Try using `Sequience`, that would allow driver not to stop every X km because we don't process routes by chunks 
* Try reusing existing decoded route geometry from cache

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteMapper.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteMapper.kt:100]: That is essentially the approach I was taking here https://github.com/mapbox/mapbox-navigation-android/pull/6582

Instead of a Stream it's an `Iterator`. I'm not opposed to introducing it, but it still requires bug fixing
``` java
public interface ReplayEventStream extends Closeable, Iterator<ReplayEventBase>
```
